File: src/getdate_next/lexer.py
"""
lexer.py - Tokenizer for natural language date expressions.

Provides tokenization of date strings into a stream of tokens
for parsing by the datetime parser.
"""

from enum import Enum, auto
from dataclasses import dataclass
from typing import Optional


class TokenType(Enum):
    NUMBER = auto()
    WORD = auto()
    MODIFIER = auto()
    TIME = auto()
    OFFSET = auto()
    ORDINAL = auto()
    DAY = auto()
    MONTH = auto()
    UNIT = auto()
    DATE_SEPARATOR = auto()
    TIME_SEPARATOR = auto()
    TIMEZONE = auto()
    COMMA = auto()
    UNIX = auto()
    AT = auto()
    EOF = auto()


@dataclass
class Token:
    type: TokenType
    value: str
    position: int


MODIFIERS = {"next", "last", "previous"}
OFFSETS = {"ago", "from", "now", "until", "since"}
UNITS = {
    "day",
    "days",
    "week",
    "weeks",
    "month",
    "months",
    "year",
    "years",
    "hour",
    "hours",
    "minute",
    "minutes",
    "second",
    "seconds",
}

DAYS = {
    "monday",
    "tuesday",
    "wednesday",
    "thursday",
    "friday",
    "saturday",
    "sunday",
    "mon",
    "tue",
    "wed",
    "thu",
    "fri",
    "sat",
    "sun",
}

MONTHS = {
    "january",
    "february",
    "march",
    "april",
    "may",
    "june",
    "july",
    "august",
    "september",
    "october",
    "november",
    "december",
    "jan",
    "feb",
    "mar",
    "apr",
    "jun",
    "jul",
    "aug",
    "sep",
    "oct",
    "nov",
    "dec",
}

TIME_WORDS = {"noon", "midday", "midnight", "night", "morning", "evening", "am", "pm", "a", "p", "of"}

ORDINALS = {
    "first",
    "second",
    "third",
    "fourth",
    "fifth",
    "sixth",
    "seventh",
    "eighth",
    "ninth",
    "tenth",
    "eleventh",
    "twelfth",
    "1st",
    "2nd",
    "3rd",
    "4th",
    "5th",
    "6th",
    "7th",
    "8th",
    "9th",
    "10th",
    "11th",
    "12th",
}

from .timezone_data import TIMEZONE_ABBREV_OFFSET

TIMEZONES = set(TIMEZONE_ABBREV_OFFSET.keys())


class TokenizerError(Exception):
    """Raised when tokenization fails."""

    pass


class Tokenizer:
    def __init__(self, buf: str):
        self.buf = buf.strip().lower()
        self.pos = 0
        self.tokens: list[Token] = []
        self.prev_token_type: Optional[TokenType] = None

    def tokenize(self) -> list[Token]:
        """Tokenize the input string."""
        while self.pos < len(self.buf):
            self._skip_whitespace()
            if self.pos >= len(self.buf):
                break

            token = self._next_token()
            if token:
                self.tokens.append(token)
                self.prev_token_type = token.type

        self.tokens.append(Token(TokenType.EOF, "", self.pos))
        return self.tokens

    def _skip_whitespace(self) -> None:
        while self.pos < len(self.buf) and self.buf[self.pos].isspace():
            self.pos += 1

    def _next_token(self) -> Optional[Token]:
        char = self.buf[self.pos]

        if char.isdigit():
            return self._tokenize_number()
        if char.isalpha():
            return self._tokenize_word()
        if char in "-/+":
            return self._tokenize_separator()
        if char in ":":
            return self._tokenize_time_separator()
        if char == ",":
            return self._tokenize_comma()

        self.pos += 1
        return None

    def _tokenize_comma(self) -> Token:
        char = self.buf[self.pos]
        self.pos += 1
        return Token(TokenType.COMMA, char, self.pos - 1)

    def _tokenize_number(self) -> Token:
        start = self.pos
        while self.pos < len(self.buf) and self.buf[self.pos].isdigit():
            self.pos += 1

        value = self.buf[start : self.pos]

        if self.pos < len(self.buf) and self.buf[self.pos] in "stndrth":
            suffix = ""
            while self.pos < len(self.buf) and self.buf[self.pos] in "stndrth":
                suffix += self.buf[self.pos]
                self.pos += 1
            value = value + suffix
            if value in ORDINALS:
                return Token(TokenType.ORDINAL, value, start)
            return Token(TokenType.NUMBER, value, start)

        if self.pos < len(self.buf):
            next_char = self.buf[self.pos].lower()
            if (
                next_char in "ap"
                or self.buf[self.pos :].startswith("am")
                or self.buf[self.pos :].startswith("pm")
            ):
                if len(value) >= 3 and len(value) <= 4 and value.isdigit():
                    return Token(TokenType.NUMBER, value, start)
                ampm = ""
                while self.pos < len(self.buf) and self.buf[self.pos].isalpha():
                    ampm += self.buf[self.pos]
                    self.pos += 1
                return Token(TokenType.TIME, value + ampm, start)

            if next_char == ":":
                time_val = value
                while self.pos < len(self.buf) and self.buf[self.pos] in "0123456789:":
                    time_val += self.buf[self.pos]
                    self.pos += 1
                if self.pos < len(self.buf) and self.buf[self.pos].lower() in "ap":
                    ampm = ""
                    while self.pos < len(self.buf) and self.buf[self.pos].isalpha():
                        ampm += self.buf[self.pos]
                        self.pos += 1
                    return Token(TokenType.TIME, time_val + ampm, start)
                return Token(TokenType.TIME, time_val, start)

        if len(value) == 10 and value.isdigit():
            return Token(TokenType.UNIX, value, start)

        return Token(TokenType.NUMBER, value, start)

    def _tokenize_word(self) -> Token:
        start = self.pos
        while self.pos < len(self.buf) and self.buf[self.pos].isalnum():
            self.pos += 1

        value = self.buf[start : self.pos]

        if value in MODIFIERS:
            return Token(TokenType.MODIFIER, value, start)
        if value in DAYS:
            return Token(TokenType.DAY, value, start)
        if value in MONTHS:
            return Token(TokenType.MONTH, value, start)
        if value in TIME_WORDS:
            return Token(TokenType.TIME, value, start)
        if value in UNITS:
            return Token(TokenType.UNIT, value, start)
        if value in OFFSETS:
            return Token(TokenType.OFFSET, value, start)
        if value in ORDINALS:
            return Token(TokenType.ORDINAL, value, start)
        if value in TIMEZONES:
            return Token(TokenType.TIMEZONE, value, start)
        if value == "at":
            return Token(TokenType.AT, value, start)

        return Token(TokenType.WORD, value, start)

    def _tokenize_separator(self) -> Token:
        char = self.buf[self.pos]
        self.pos += 1

        if char in "+-" and self.prev_token_type == TokenType.TIME:
            tz_val = char
            while self.pos < len(self.buf) and (self.buf[self.pos].isdigit() or self.buf[self.pos] == ":"):
                tz_val += self.buf[self.pos]
                self.pos += 1
            return Token(TokenType.TIMEZONE, tz_val, self.pos - len(tz_val))

        return Token(TokenType.DATE_SEPARATOR, char, self.pos - 1)

    def _tokenize_time_separator(self) -> Token:
        char = self.buf[self.pos]
        self.pos += 1
        return Token(TokenType.TIME_SEPARATOR, char, self.pos - 1)


def tokenize(buf: str) -> list[Token]:
    """
    Tokenize a date expression string.

    Args:
        buf: Date expression string

    Returns:
        List of tokens
    """
    if not buf or not isinstance(buf, str):
        return []

    tokenizer = Tokenizer(buf)
    return tokenizer.tokenize()


if __name__ == "__main__":
    test_cases = [
        "next friday 2pm",
        "2nd wednesday of march 2026",
        "+2 days",
        "3 days ago",
        "2026-03-06T14:30:00Z",
        "tomorrow noon",
        "next week",
        "final friday of march",
    ]

    for tc in test_cases:
        tokens = tokenize(tc)
        print(f"{tc!r}:")
        for tok in tokens:
            if tok.type == TokenType.EOF:
                break
            print(f"  {tok.type.name:15} {tok.value!r:20}")
        print()

File: php/src/Lexer.php
<?php

declare(strict_types=1);

namespace GetdateNext;

use UnitEnum;

enum TokenType: string
{
    case NUMBER = 'NUMBER';
    case WORD = 'WORD';
    case MODIFIER = 'MODIFIER';
    case TIME = 'TIME';
    case OFFSET = 'OFFSET';
    case ORDINAL = 'ORDINAL';
    case DAY = 'DAY';
    case MONTH = 'MONTH';
    case UNIT = 'UNIT';
    case DATE_SEPARATOR = 'DATE_SEPARATOR';
    case TIME_SEPARATOR = 'TIME_SEPARATOR';
    case TIMEZONE = 'TIMEZONE';
    case COMMA = 'COMMA';
    case UNIX = 'UNIX';
    case AT = 'AT';
    case EOF = 'EOF';
}

class Token
{
    public function __construct(
        public readonly TokenType $type,
        public readonly string $value,
        public readonly int $position,
    ) {
    }
}

class Lexer
{
    private const MODIFIERS = ['next', 'last', 'previous'];
    private const OFFSETS = ['ago', 'from', 'now', 'until', 'since'];
    private const UNITS = ['day', 'days', 'week', 'weeks', 'month', 'months', 'year', 'years', 'hour', 'hours', 'minute', 'minutes', 'second', 'seconds'];
    private const DAYS = ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday', 'mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'];
    private const MONTHS = ['january', 'february', 'march', 'april', 'may', 'june', 'july', 'august', 'september', 'october', 'november', 'december', 'jan', 'feb', 'mar', 'apr', 'jun', 'jul', 'aug', 'sep', 'oct', 'nov', 'dec'];
    private const TIME_WORDS = ['noon', 'midday', 'midnight', 'night', 'morning', 'evening', 'am', 'pm', 'a', 'p', 'of'];
    private const ORDINALS = ['first', 'second', 'third', 'fourth', 'fifth', 'sixth', 'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth', '1st', '2nd', '3rd', '4th', '5th', '6th', '7th', '8th', '9th', '10th', '11th', '12th'];
    private const TIMEZONES = ['est', 'cst', 'mst', 'pst', 'edt', 'cdt', 'mdt', 'pdt', 'gmt', 'utc', 'z', 'bst', 'cet', 'cest', 'jst', 'aest', 'aedt', 'nzst', 'nzdt'];

    private string $buf;
    private int $pos = 0;
    private array $tokens = [];
    private ?TokenType $prevTokenType = null;

    public function tokenize(string $buf): array
    {
        $this->buf = strtolower(trim($buf));
        $this->pos = 0;
        $this->tokens = [];
        $this->prevTokenType = null;

        while ($this->pos < strlen($this->buf)) {
            $this->skipWhitespace();
            if ($this->pos >= strlen($this->buf)) {
                break;
            }

            $token = $this->nextToken();
            if ($token !== null) {
                $this->tokens[] = $token;
                $this->prevTokenType = $token->type;
            }
        }

        $this->tokens[] = new Token(TokenType::EOF, '', $this->pos);
        return $this->tokens;
    }

    private function skipWhitespace(): void
    {
        while ($this->pos < strlen($this->buf) && ctype_space($this->buf[$this->pos])) {
            $this->pos++;
        }
    }

    private function nextToken(): ?Token
    {
        if ($this->pos >= strlen($this->buf)) {
            return null;
        }

        $char = $this->buf[$this->pos];

        if (ctype_digit($char)) {
            return $this->tokenizeNumber();
        }
        if (ctype_alpha($char)) {
            $start = $this->pos;
            // Check for ISO 8601 T separator
            if (strtolower($char) === 't' && $this->pos + 1 < strlen($this->buf) && ctype_digit($this->buf[$this->pos + 1])) {
                $this->pos++;
                return new Token(TokenType::TIME_SEPARATOR, 'T', $start);
            }
            return $this->tokenizeWord();
        }
        if (in_array($char, ['-', '+', '/'], true)) {
            return $this->tokenizeSeparator();
        }
        if ($char === ':') {
            return $this->tokenizeTimeSeparator();
        }
        if ($char === ',') {
            return $this->tokenizeComma();
        }

        $this->pos++;
        return null;
    }

    private function tokenizeNumber(): Token
    {
        $start = $this->pos;
        while ($this->pos < strlen($this->buf) && ctype_digit($this->buf[$this->pos])) {
            $this->pos++;
        }

        $value = substr($this->buf, $start, $this->pos - $start);

        if ($this->pos < strlen($this->buf) && in_array($this->buf[$this->pos], ['s', 'n', 'd', 'r', 'h'], true)) {
            $suffix = '';
            while ($this->pos < strlen($this->buf) && in_array($this->buf[$this->pos], ['s', 't', 'n', 'd', 'r', 't', 'h'], true)) {
                $suffix .= $this->buf[$this->pos];
                $this->pos++;
            }
            $value = $value . $suffix;
            if (in_array($value, self::ORDINALS, true)) {
                return new Token(TokenType::ORDINAL, $value, $start);
            }
            return new Token(TokenType::NUMBER, $value, $start);
        }

        if ($this->pos < strlen($this->buf)) {
            $nextChar = strtolower($this->buf[$this->pos]);
            if (in_array($nextChar, ['a', 'p'], true) || str_starts_with(substr($this->buf, $this->pos), 'am') || str_starts_with(substr($this->buf, $this->pos), 'pm')) {
                $ampm = '';
                while ($this->pos < strlen($this->buf) && ctype_alpha($this->buf[$this->pos])) {
                    $ampm .= $this->buf[$this->pos];
                    $this->pos++;
                }
                return new Token(TokenType::TIME, $value . $ampm, $start);
            }

            if ($nextChar === ':') {
                $timeVal = $value;
                while ($this->pos < strlen($this->buf) && (ctype_digit($this->buf[$this->pos]) || $this->buf[$this->pos] === ':')) {
                    $timeVal .= $this->buf[$this->pos];
                    $this->pos++;
                }
                if ($this->pos < strlen($this->buf) && in_array(strtolower($this->buf[$this->pos]), ['a', 'p'], true)) {
                    $ampm = '';
                    while ($this->pos < strlen($this->buf) && ctype_alpha($this->buf[$this->pos])) {
                        $ampm .= $this->buf[$this->pos];
                        $this->pos++;
                    }
                    return new Token(TokenType::TIME, $timeVal . $ampm, $start);
                }
                return new Token(TokenType::TIME, $timeVal, $start);
            }
        }

        if (strlen($value) === 10 && ctype_digit($value)) {
            return new Token(TokenType::UNIX, $value, $start);
        }

        return new Token(TokenType::NUMBER, $value, $start);
    }

    private function tokenizeWord(): Token
    {
        $start = $this->pos;
        while ($this->pos < strlen($this->buf) && ctype_alnum($this->buf[$this->pos])) {
            $this->pos++;
        }

        $value = substr($this->buf, $start, $this->pos - $start);

        if (in_array($value, self::MODIFIERS, true)) {
            return new Token(TokenType::MODIFIER, $value, $start);
        }
        if (in_array($value, self::DAYS, true)) {
            return new Token(TokenType::DAY, $value, $start);
        }
        if (in_array($value, self::MONTHS, true)) {
            return new Token(TokenType::MONTH, $value, $start);
        }
        if (in_array($value, self::TIME_WORDS, true)) {
            return new Token(TokenType::TIME, $value, $start);
        }
        if (in_array($value, self::UNITS, true)) {
            return new Token(TokenType::UNIT, $value, $start);
        }
        if (in_array($value, self::OFFSETS, true)) {
            return new Token(TokenType::OFFSET, $value, $start);
        }
        if (in_array($value, self::ORDINALS, true)) {
            return new Token(TokenType::ORDINAL, $value, $start);
        }
        if (in_array($value, self::TIMEZONES, true)) {
            return new Token(TokenType::TIMEZONE, $value, $start);
        }
        if ($value === 'at') {
            return new Token(TokenType::AT, $value, $start);
        }

        return new Token(TokenType::WORD, $value, $start);
    }

    private function tokenizeSeparator(): Token
    {
        $char = $this->buf[$this->pos];
        $start = $this->pos;
        $this->pos++;

        if (in_array($char, ['+', '-'], true) && $this->prevTokenType === TokenType::TIME) {
            $tzVal = $char;
            while ($this->pos < strlen($this->buf) && (ctype_digit($this->buf[$this->pos]) || $this->buf[$this->pos] === ':')) {
                $tzVal .= $this->buf[$this->pos];
                $this->pos++;
            }
            return new Token(TokenType::TIMEZONE, $tzVal, $start);
        }

        return new Token(TokenType::DATE_SEPARATOR, $char, $start);
    }

    private function tokenizeTimeSeparator(): Token
    {
        $char = $this->buf[$this->pos];
        $this->pos++;
        return new Token(TokenType::TIME_SEPARATOR, $char, $this->pos - 1);
    }

    private function tokenizeComma(): Token
    {
        $char = $this->buf[$this->pos];
        $this->pos++;
        return new Token(TokenType::COMMA, $char, $this->pos - 1);
    }
}